File: lib/parser.rb
require 'ast/builder'

class Parser
  def initialize(builder)
    self.builder = builder
  end

  def parse(string)
    self.string = string
    parse_everything
  end

  def self.parse(string)
    self.new(AST::Builder.new).parse(string)
  end

  private

  attr_accessor :builder
  attr_reader :string

  def string=(string)
    @string = string.strip
  end

  def parse_everything
    expression = parse_sequence
    read %r{\z}

    expression
  end

  def parse_sequence
    expressions = []
    expressions << parse_expression until can_read? %r{\)|\z}

    if expressions.length == 1
      expressions.first
    else
      builder.build_sequence(expressions)
    end
  end

  def parse_expression
    if can_read? %r{\(}
      parse_brackets
    elsif can_read? %r{\$}
      parse_variable
    elsif can_read? %r{[^\p{Blank}()]+}
      parse_keyword
    else
      complain
    end
  end

  def parse_brackets
    read %r{\(}
    expression = parse_sequence
    read %r{\)}

    expression
  end

  def parse_variable
    read %r{\$}
    builder.build_variable(read_name)
  end

  def parse_keyword
    builder.build_keyword(read_name)
  end

  def read_name
    read %r{[^\p{Blank}()]+}
  end

  def can_read?(pattern)
    !try_match(pattern).nil?
  end

  def read(pattern)
    match = try_match(pattern) || complain(pattern)
    self.string = match.post_match
    match.to_s
  end

  def try_match(pattern)
    /\A#{pattern}/.match(string)
  end

  def complain(expected = nil)
    complaint = "unexpected #{string.slice(0)}"
    complaint << ", expected #{expected.inspect}" if expected

    raise complaint
  end
end

File: src/parser.js
export default class {
  constructor(builder) {
    this.builder = builder;
  }

  parse(string) {
    this.string = string;
    return this.parseEverything();
  }

  get string() {
    return this._string;
  }

  set string(string) {
    this._string = string.trim();
  }

  parseEverything() {
    const expression = this.parseSequence();
    this.read(/$/);

    return expression;
  }

  parseSequence() {
    const expressions = [];

    while (!this.canRead(/\)|$/)) {
      expressions.push(this.parseExpression());
    }

    if (expressions.length === 1) {
      return expressions[0];
    } else {
      return this.builder.buildSequence(expressions);
    }
  }

  parseExpression() {
    if (this.canRead(/\(/)) {
      return this.parseBrackets();
    } else if (this.canRead(/_/)) {
      return this.parseVariable();
    } else if (this.canRead(/[^\s()]+/)) {
      return this.parseKeyword();
    } else {
      this.complain();
    }
  }

  parseBrackets() {
    this.read(/\(/);
    const expression = this.parseSequence();
    this.read(/\)/);

    return expression;
  }

  parseVariable() {
    this.read(/_/);
    return this.builder.buildVariable(this.readName());
  }

  parseKeyword() {
    return this.builder.buildKeyword(this.readName());
  }

  readName() {
    return this.read(/[^\s()]+/);
  }

  canRead(pattern) {
    return this.tryMatch(pattern) !== null;
  }

  read(pattern) {
    const match = this.tryMatch(pattern);
    if (match === null) {
      this.complain(pattern);
    }

    this.string = match[2];

    return match[1];
  }

  tryMatch(pattern) {
    return this.string.match(new RegExp('^(' + pattern.source + ')(.*)$'));
  }

  complain(expected) {
    let complaint = 'unexpected ';

    if (this.string.length === 0) {
      complaint += 'end of string';
    } else {
      complaint += '"' + this.string.charAt(0) + '"';
    }

    if (expected !== undefined) {
      complaint += ', expected ' + expected;
    }

    throw new Error(complaint);
  }
}
